clippy_lints/src/methods/search_is_some.rs

```diff
@@ -146,3 +171,187 @@ pub(super) fn check<'tcx>(
+    fn consume(&mut self, _: &PlaceWithHirId<'tcx>, _: HirId) {}
+
+    fn borrow(&mut self, cmt: &PlaceWithHirId<'tcx>, _: HirId, _: ty::BorrowKind) {
+        if let PlaceBase::Local(id) = cmt.place.base {
+            let map = self.cx.tcx.hir();
+            let ident_str = map.name(id).to_string();
+            let span = map.span(cmt.hir_id);
+            let start_span = Span::new(self.next_pos, span.lo(), span.ctxt());
+            let mut start_snip = snippet_with_applicability(self.cx, start_span, "..", &mut self.applicability);
+
+            if cmt.place.projections.is_empty() {
+                // handle item without any projection, that needs an explicit borrowing
+                // i.e.: suggest `&x` instead of `x`
+                self.suggestion_start.push_str(&format!("{}&{}", start_snip, ident_str));
+            } else {
+                // cases where a parent call is using the item
+                // i.e.: suggest `.contains(&x)` for `.find(|x| [1, 2, 3].contains(x)).is_none()`
+                if let Some(parent_expr) = get_parent_expr_for_hir(self.cx, cmt.hir_id) {
+                    if let ExprKind::Call(_, call_args) | ExprKind::MethodCall(_, _, call_args, _) = parent_expr.kind {
+                        let expr = self.cx.tcx.hir().expect_expr(cmt.hir_id);
+                        let arg_ty_kind = self.cx.typeck_results().expr_ty(expr).kind();
+
+                        // Note: this should always be true, as `find` only gives us a reference which are not mutable
+                        if matches!(arg_ty_kind, ty::Ref(_, _, Mutability::Not)) {
+                            let start_span = Span::new(self.next_pos, span.lo(), span.ctxt());
+                            let start_snip =
+                                snippet_with_applicability(self.cx, start_span, "..", &mut self.applicability);
+                            // do not suggest ampersand if the ident is the method caller
+                            let ident_sugg = if !call_args.is_empty() && call_args[0].hir_id == cmt.hir_id {
+                                format!("{}{}", start_snip, ident_str)
+                            } else {
+                                format!("{}&{}", start_snip, ident_str)
+                            };
+                            self.suggestion_start.push_str(&ident_sugg);
+                            self.next_pos = span.hi();
+                        } else {
+                            self.applicability = Applicability::Unspecified;
+                        }
+                        return;
+                    }
+                }
+
+                // handle item projections by removing one explicit deref
+                // i.e.: suggest `*x` instead of `**x`
+                let mut replacement_str = ident_str;
+
+                // handle index projection first
+                let index_handled = cmt.place.projections.iter().any(|proj| match proj.kind {
```

Comment:
Here I found another edge case when the value is dereferenced before the index expression, like this:

```
let _ = v.iter().find(|x| (**x)[0] == 9).is_some();
```
This results in the following projections: `[Projection { ty: &[i32; 4], kind: Deref }, Projection { ty: [i32; 4], kind: Deref }, Projection { ty: i32, kind: Index }]`. Which leads to the suggestion: `any(|x| (**x)[0] == 9)`

----

It seems like the `ExprUseVisitor` is resolving all projections at once. The order of the projections seems to by in sync with what's happening in the code.

Based on this, I'm starting to wonder if it would be simpler to use a normal `Visitor` instead, with [`path_to_local`](https://github.com/rust-lang/rust-clippy/blob/2316f4da83a3952abca808f1f1b2c7ea0a9ee401/clippy_utils/src/lib.rs#L497). Then we could explicitly ask for the parent expression and only get one level above, instead of this where we get all projections at once. I'm not sure on this, though. And it would also mean some more work for you. :sweat_smile: